Add Kubeadm-DinD-Cluster based kubetest deployer for IPv6 CI

leblancd opened the pull request:
This change adds a Kubeadm-DinD-Cluster based kubetest deployer that will be used to create a
CI for Kubernetes IPv6-only functionality.

The Prow job config for this CI will be added via a separate PR (see PR #9640).

This CI test uses deployment scripts from the kubernetes-sigs/kubeadm-dind-cluster
github (K-D-C) repo to set up a containerized (Docker-in-Docker-in-Docker), multi-node Kubernetes
cluster that can be run Prow container.

The K-D-C deployment scripts provide some built-in functionality that is important for testing IPv6-only support (see https://github.com/leblancd/kube-v6 for details):
- Bridge CNI and host local IPAM plugins are loaded (both plugins support IPv6)
- Static routes are provided from each node to pod networks on other nodes.
- When the scripts are run in IPv6 mode, the following is configured for the cluster:
     * IPv6 node addresses and service subnets
     * IPv6 CNI configuration for pods
     * IPv6 DNS nameserver in kubelet startup config
     * Bind9 container for DNS64 functionality
     * Tayga container for NAT64 functionality

The End-to-End test suite that is used to test functionality is described here:
    https://github.com/CiscoSystems/kube-v6-test

I've tested this CI by running its Prow container locally (via 'docker run ...'), [sample results and logs can be found here](https://k8s-gubernator.appspot.com/build/my-kubernetes-jenkins/pull/65951/pull-kubernetes-e2e-kubeadm-dind-ipv6/20180813-230929--115b21b422a0b875-197/)

Note: With this current change set, kubeadm-dind-cluster always does its own build of Kubernetes.
As a potential future CI performance improvement, this can be changed to have kubeadm-dind-cluster use a shared build from GCS.

krzyzacy commented:
/assign @Q-Lee

fejta commented:
/assign @Q-Lee

BenTheElder commented on prow/config.yaml:
```diff
@@ -1302,6 +1302,62 @@ presubmits:
+          secret:
+            secretName: ssh-key-secret
+            defaultMode: 0400
+        - name: cache-ssd
```
do not use this.

BenTheElder commented on prow/config.yaml:
```diff
@@ -1302,6 +1302,62 @@ presubmits:
+            mountPath: /etc/ssh-key-secret
+            readOnly: true
+          - name: cache-ssd
+            mountPath: /root/.cache
```
please remove this mount and the containerPort, I've just finished eliminating this.

leblancd replied:
Will remove these, thanks.

krzyzacy commented:
/ok-to-test

Q-Lee commented on images/gce-dind/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+    apt-get clean
+
+# Install kubeadm-dind-cluster scripts
+RUN cd /root && \
```
Rather than baking it in, we should pull git repos at point of use, so it's easier to change versions.

BenTheElder replied:
we can do this with a second --repo bootstrap flag on the job instead of baking it into an image

leblancd replied:
Good idea, thanks. Will change this.

leblancd replied:
Done. Added a --repo flag for the kubeadm-dind-cluster repo.

Q-Lee commented on images/gce-dind/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+    git clone -b k8s_ipv6_ci https://github.com/leblancd/kubeadm-dind-cluster.git
+
+# Install docker-machine
+RUN curl -L https://github.com/docker/machine/releases/download/v0.13.0/docker-machine-`uname -s`-`uname -m` >/tmp/docker-machine && install /tmp/docker-machine /usr/local/bin/docker-machine
```
Why do we need docker machine, and shouldn't this be managed by apt-get?

leblancd replied:
The kubeadm-dind-cluster scripts use docker-machine to both create the instance on GCE, and to create a connection to the docker daemon running on that instance. Once the docker-machine connection is set up, the docker client in the Prow container will use the docker daemon in the GCE instance, e.g. for commands like 'docker run ...', 'docker ps ...', etc.

The [recommended way to install docker-machine](https://docs.docker.com/machine/install-machine/) is to curl the binary.

Q-Lee commented on kubetest/gce_dind.go:
```diff
@@ -0,0 +1,585 @@
+// to create a containerized, Docker-in-Docker (DinD) based Kubernetes cluster
+// that is running in a remote GCE instance.
+
+package main
```
We're trying to avoid putting new kubetest implementations into the main package. It's been a non-trivial effort to begin teasing them apart.

leblancd replied:
Will move this out of the main package.

leblancd replied:
Done. Moved gce_dind.go to its own "gcedind" package (under kubetest).

leblancd commented:
@Q-Lee : I've addressed the comments in the first round of review. Any chance you can take a look?

Q-Lee commented on kubetest/gcedind/gce_dind.go:
```diff
@@ -0,0 +1,622 @@
+
+var (
+	// gce-dind specific test config flags.
+	gceDinDScriptsPath = flag.String("gce-dind-scripts-path", "",
```
Let's try to avoid defining flags in packages. These are effectively globals.

leblancd replied:
I eliminated most flags, and removed remainder to main.go

Q-Lee commented on kubetest/gcedind/gce_dind.go:
```diff
@@ -0,0 +1,622 @@
+		"(gce-dind only) Indicates whether cluster should run in IPv6-only mode.")
+	gceDinDUpTimeout = flag.Duration("gce-dind-up-timeout", 45*time.Minute,
+		"(gce-dind only) Time limit between starting a cluster and making a successful call to the Kubernetes API.")
+	gceDinDNumNodes = flag.Int("gce-dind-num-nodes", 4,
```
Again, let's try to avoid flag explosion.

Either use the existing gcp-nodes flag, or [better yet] make a --num-nodes flag. It can default to 4, and [if set] deployment specific flag variables can take their value from this.

leblancd replied:
Created a common --num-nodes flag.

leblancd commented on kubetest/gcedind/gce_dind.go:
```diff
@@ -0,0 +1,622 @@
+		"(gce-dind only) Indicates whether hyperkube should be built.")
+	gceDinDIPv6Only = flag.Bool("gce-dind-ipv6-only", false,
+		"(gce-dind only) Indicates whether cluster should run in IPv6-only mode.")
+	gceDinDUpTimeout = flag.Duration("gce-dind-up-timeout", 45*time.Minute,
```
It takes roughly 30 minutes. So perhaps the 45 minute timeout is a tad too tight. Here are some [recent manual runs](https://k8s-gubernator.appspot.com/builds/my-kubernetes-jenkins/pull/61987/pull-kubernetes-e2e-gce-dind-ipv6/) for reference.

Q-Lee commented on kubetest/gcedind/gce_dind.go:
```diff
@@ -0,0 +1,622 @@
+
+// saveInstanceLogs collects a serial log, service logs, and docker state
+// from the GCE instance, and saves the logs in a local artifacts directory.
+func (d *Deployer) saveInstanceLogs(artifactsDir string) error {
```
This functionality is more general than the dind tests. Could you move this code into another package, so we can later use it to break the general gce e2e test dependency on the log recovery bash script in k/k?

leblancd replied:
I'll take a look at how this can be factored out. Just to make sure we're on the same page...
Is the functionality that you're discussing moving out just for the GCE instance, or for both GCE instance + Master/Worker nodes (i.e. functionality that would apply to any setup that has master/worker nodes that are reachable via 'docker exec <container> <cmd>')?

If I move the GCE instance logging functionality to a new (general purpose) package, it would have to be a little bit flexible in allowing the caller/user to provide lists of what to collect. For example, in my specific GCE DinD case, the one GCE instance won't be doing kube-api, kube-controller, kube-proxy, etc., but in other GCE cases, a GCE master instance will be running these system pods. (Probably not a big deal, just argument passing).

For the master/worker log collection, I'm using 'docker exec <container> <cmd>' to collect logs (this works because of the docker-machine connection to the instance/VM), but this might also work for other cases, e.g. the local DinD case, I believe.

Q-Lee replied:
We should probably take those lists as formal parameters anyway, or at least from member fields.

I just want the log scraping for a gce node moved out. The bash script we currently use in the e2e tests is functionally similar to this, but has deep import dependencies on the k/k/cluster dir. For pods, I suspect we'll be able to just scrape the logs from the mirror pods via the k8s api.

I'll read through this again tomorrow and check that I'm making sense.

leblancd replied:
I took a look at how the GCE instance log collecting can be factored out separately in a general-purpose way that will work for other GCE-based CIs. I think this is a great idea, but I'm wondering if this should be done as a separate, followup change. My thought process is as follows: I started by looking at kubetest/dump.go (used in the kops-based CI tests) as a model for what might be needed. (The dump.go relies on the user providing an SSH client factory for creating client connections for each node VM. Perhaps GCE-base CIs could re-use this dump.go framework with a standard SSH client, since 'gcloud compute ssh ...' is supposedly a wrapper around ssh???) Using dump.go as a model, this framework uses 'kubectl get nodes' and then calls the d.dumpNode() method to collect the logs for each node in turn. So let's say that I do go ahead and add a new "gcedump" package to provide similar general-purpose functionality for GCE-based CI. Most of the logic in this new package wouldn't get exercised in the gce-DinD case. For example, I would have to tell the "gcedump" to ignore the 'kubectl get nodes' output (since kubernetes nodes are containers, not VMs, in gce-DinD), and I'd have to tell it to also collect logs for the GCE instance in this special case. So in order to test this new "gcedump" framework, I would also have to test it on an existing, more typical GCE-based CI test job. This would be development work that's really outside the scope of what I'm trying to add for GCE-DinD. Or I could just add a package that only collects logs for a single GCE VM, but ultimately I think we want something that does the log collection for all nodes on GCE-based tests. Your thoughts?

Q-Lee commented on kubetest/gcedind/gce_dind.go:
```diff
@@ -0,0 +1,622 @@
+	// each individual log file from the container node.
+	for _, logFile := range logFiles {
+		log.Printf("Saving %s", logFile)
+		cmd := fmt.Sprintf("find %s -name %s -print", nodeLogDir, logFile)
```
@ixdy how much do we care about injections in kubetest? Eventually these logFiles might become parameterized.

leblancd commented on kubetest/gcedind/gce_dind.go:
```diff
@@ -0,0 +1,622 @@
+// detectNodeContainers creates a list of names for either all master or all
+// worker node containers. It does this by running 'docker ps' and searching
+// for container names that begin with a specified name prefix.
+func (d *Deployer) detectNodeContainers(namePrefix string) ([]string, error) {
```
Maybe you meant listing nodes? Unfortunately, listing pods won't work in this case since the master/worker node containers are running as "plain" (non-k8s) docker containers on the GCE instance. Listing nodes via kubectl would work, and might be safer, although the downside would be that the node-listing approach couldn't be used if we ever enhance the log collection to include gathering stuff from the tayga and bind9 (non-k8s) containers. Since we're not collecting tayga/bind9 info yet, I could switch this to use kubectl listing of nodes for now.

Q-Lee replied:
I'm still not sold on making the tayga and bind9 tests blocking CI, because it creates an internet dependency, and it seems like it's testing your configuration more than kubernetes.

That said, why would it not be able to pull out tayga and bind9 logs? Are you running them in the "top" level of docker?

leblancd replied:
There is one aspect of Kubernetes for which the external ping tests do add *some* value: When pinging from a pod to an external address, kube-proxy/ip6tables needs to masquerade (SNAT) the pod addresses, since the pod subnet is using private/ULA space (pod addresses aren't meaningful externally). Granted, the SNAT'd addresses will only exist on the link between the kube node and the tayga container, since tayga will swap in an IPv4 source address through NAT64.
Yes, tayga and bind9 are running on the "top" or "outer" docker layer (along with the master and worker nodes).

leblancd commented on kubetest/gcedind/gce_dind.go:
```diff
@@ -0,0 +1,622 @@
+// and cleans up the docker-machine state.
+func (d *Deployer) Down() error {
+	// Check whether expected docker-machine exists
+	o, err := d.getOutput("docker-machine ls")
```
I couldn't find any docker-machine Golang SDK. However, I did find this:
 [go-docker-machine](https://github.com/AkihiroSuda/go-docker-machine) 
but this simply adds some logic around running 'docker-machine \<cmd\> \<args\>' via os/exec, similar to what's done here. Also, the go-docker-machine README says it's only been tested up to docker-machine 0.6.0 (we're using 0.12.1), and it hasn't been updated in a couple of years.

leblancd commented on kubetest/gcedind/gce_dind.go:
```diff
@@ -0,0 +1,622 @@
+		"(gce-dind only) Indicates whether kubeadm should be built.")
+	gceDinDBuildHyperkube = flag.Bool("gce-dind-build-hyperkube", false,
+		"(gce-dind only) Indicates whether hyperkube should be built.")
+	gceDinDIPv6Only = flag.Bool("gce-dind-ipv6-only", false,
```
Makes sense.

leblancd replied:
Done.

Q-Lee commented on kubetest/gcedind/gce_dind.go:
```diff
@@ -0,0 +1,622 @@
+		}
+	}
+
+	// Copy log files for kube system pods that are running in this node
```
>Use 'kubectl logs ...' The disadvantage: won't include logs for dead pods, which we really want.

That's a good point. Although I wonder if we could alter the GC's aggressiveness, but that's neither here nor there.

Q-Lee commented:
@krzyzacy could you review the boskos behavior required here?

krzyzacy commented:
So this deployer is expected to use boskos to manage gcp projects?

leblancd commented:
/test pull-test-infra-lint

leblancd commented:
/test pull-test-infra-bazel

leblancd commented:
@Q-Lee I tested the prow container using 'docker run ...'
Here's a [gist with instructions on how I tested](https://gist.github.com/leblancd/62ad517a01049ba9c614af4b500077f3)

BenTheElder commented:
@Q-Lee we can construct a prow job but I don't think it matters, my concerns are:
- this needs to be maintained by someone, preferably someone(s)
- we need to make sure the resources don't leak (@krzyzacy seems to have signed off on that)
- there might be overlap with other dind work.

since this is just a CI job we can create it, let it fail, and follow up with necessary fixes, assuming the above are accceptable.

leblancd commented:
@Q-Lee At some point, I should probably add a corresponding post-commit/periodic test job. The job config wouldn't look too different. Should I wait on adding this for now?

BenTheElder commented on prow/config.yaml:
```diff
@@ -1937,6 +1937,38 @@ presubmits:
+    agent: kubernetes
+    context: pull-kubernetes-e2e-gce-dind-ipv6
+    max_concurrency: 8
+    skip_report: true
```
we should probably explicitly mark this `always_run: false`

BenTheElder replied:
nevermind, run_if_changed handles that :^)

BenTheElder commented on prow/config.yaml:
```diff
@@ -1937,6 +1937,38 @@ presubmits:
+    context: pull-kubernetes-e2e-gce-dind-ipv6
+    max_concurrency: 8
+    skip_report: true
+    run_if_changed: '^(cmd/kubeadm|build/debs).*$'
```
oh I see. hmm, that seems fine, ignore the comment about always run /shrug

leblancd commented:
@Q-Lee , @BenTheElder - I've added a post-commit periodic job for this. The post-commit periodic job should show up on the sig-network testgrid dashboard, if I made the changes correctly. Is there any more I need to do for this PR?

BenTheElder commented:
/assign @MrHohn

MrHohn commented:
/cc @grayluck

MrHohn commented on jobs/config.json:
```diff
@@ -12381,6 +12399,25 @@
+    "args": [
+      "--cluster=",
+      "--deployment=gce-dind",
+      "--gce-dind-image=rpothier/kubeadm-dind-cluster:k8s-1.10",
```
Does it mean the latest k8s build from the 1.10 branch? Is it possible to maintain the image on gcr.io?

rpothier replied:
This refers to the branch of kubeadm-dind-cluster. This image will be replaced by a stable branch of Mirantis/kubeadm-dind-cluster when it is available. See https://github.com/Mirantis/kubeadm-dind-cluster/issues/115

MrHohn replied:
> Once the Mirantis stable branch is created, this --gce-dind-image argument will go away

Thanks for the explanation (https://github.com/Mirantis/kubeadm-dind-cluster/issues/115 is helpful). Seems fine to keep this for now.

MrHohn commented on prow/config.yaml:
```diff
@@ -1937,6 +1937,38 @@ presubmits:
         resources:
           requests:
             memory: "6Gi"
+  - name: pull-kubernetes-e2e-gce-dind-ipv6
```
Sorry couple dumb questions: how robust will this job be? does it block PR? why is post-commit jobs not enough?

leblancd replied:
Good question. I'd have to look for guidance from the community on whether this should eventually (after some soak time) be gating/blocking, and whether a post-commit job is good enough. Our goal is for the tests to bring IPv6 to Beta status (I don't know if there is a standard for that), and to detect possible future regression bugs with a "reasonable" amount of time (maybe a 6-hour periodic is good enough for that?) In my estimation, a non-blocking pre-commit and every-6-hour post-commit would be good.

MrHohn replied:
> Our goal is for the tests to bring IPv6 to Beta status

Not sure the criteria either, maybe having all IPv6 CIs green on the release branch is sufficient?

Don't have strong opinion here, my thought on non-blocking pre-commit is that people might not care whether it fails or not if it doesn't block their PRs being merged. Likely the post-commit will still be the only place we get the failing signal :)

MrHohn commented:
cc @satyasm

MrHohn commented:
/lgtm

leblancd commented:
@MrHohn, @Q-Lee: 
I've made a couple of minor changes:
(1) The tests now use a stable branch of the Mirantis K-D-C that was created per [K-D-C issue # 115](https://github.com/Mirantis/kubeadm-dind-cluster/issues/115). What this means for this CI test is that if we need to make any changes in the future to K-D-C to resolve any issues with this CI, we won't have to make any test-infra changes.  (Instead, we'll make changes to K-D-C master, bump up the stable branch, and this CI test will pick up the changes).
(2) The test coverage has been broadened from 24 test cases to 32 test cases by modifying the ginkgo include/skip terms.

By the way, I don't know if there's a better way to handle how the Prow images are updated, but as it is implemented now (see the [Prow image config](https://github.com/kubernetes/test-infra/pull/7529/files#diff-406185368ba7839d1459d3d51424f104R2236), this test will definitely fail when it's first merged until the following happens:
- This PR gets merged (adding GCE-DinD awareness to test-infra)
- An up-to-date "e2e-gce-dind" image is made by running make in test-infra/images/gce-dind, to include the new changes.
- Someone with write permission to gcr.io/k8s-testimages registry pushes the new image.
- A second test-infra PR is created and merged to update the Prow image pointers for this CI to have it pull the updated image.
Is there a better way to do this?

Thanks!
Dane

MrHohn commented on jobs/config.json:
```diff
@@ -2762,6 +2762,23 @@
+      "--ip-mode=ipv6",
+      "--num-nodes=2",
+      "--provider=gce",
+      "--test_args=--ginkgo.focus=Networking|Services --ginkgo.skip=IPv4|DNS|Networking-Performance|Federation|functioning\\sNodePort|preserve\\ssource\\spod|affinity\\swork|affinity\\sfor --num-nodes=2 --minStartupPods=8",
```
I see some of the common tests are skipped (e.g. DNS). Is it that the component is not compatible with ipv6 or the test isn't written in a compatible way? Should we not skip them so folks are aware it needs to be fixed?

leblancd replied:
Our test coverage is described here: [IPv6 CI test coverage](https://github.com/CiscoSystems/kube-v6-test#included-test-cases). The tables on that page describe tests that are included now, failing tests that are being investigated, tests that don't make sense for our setup, and wish list tests.
As described in the "Failing Tests that are Being Investigated"](https://github.com/CiscoSystems/kube-v6-test#included-test-cases), there's an Issue filed for the DNS test coverage [k/k/ issue # 62883](https://github.com/kubernetes/kubernetes/issues/62883). A couple of mitigating factors to consider:
- Very basic kube-dns functionality is tested since the Mirantis K-D-C checks on this pod coming up (so readiness/liveness probes have to be working) before declaring the cluster up.
- Issue 62883 involves the lack of handling for PTR DNS messages in kube-dns.
- Kube-DNS will soon be replaced by coreDNS. coreDNS tests the PTR functionality, and it works there. The coreDNS folks will add IPv6 test cases with coreDNS based on the GCE-DinD framework being added here.

There are 4 failing test cases involving ESIPP (e.g. "Services [It] should be able to switch session affinity for LoadBalancer service with ESIPP on [Slow] [DisabledForLargeClusters]") that just showed up on our radar. These are disabled in our test cases, but I haven't updated our test coverage github page, nor root cause the problem. I saw your name mentioned in some ESIPP-related PRs. Do you know what these are, and whether these would be appropriate for our setup? Would this work with a GCE load balancer setup?

MrHohn replied:
Thanks, good to know the coverage is documented in detail.

ESIPP means [External Source IP Preservation](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/network/external-lb-source-ip-preservation.md). This feature requires support in both of kube-proxy and cloud provider implementation of service controller. While it should work with a GCE load balancer setup, I'm not sure whether load balancer test works on a DinD cluster. I saw the `ServiceLoadBalancer` test listed in the Wish List, guessing they might be suffering similar issue?

MrHohn commented:
> (1) The tests now use a stable branch of the Mirantis K-D-C that was created per K-D-C issue # 115. What this means for this CI test is that if we need to make any changes in the future to K-D-C to resolve any issues with this CI, we won't have to make any test-infra changes. (Instead, we'll make changes to K-D-C master, bump up the stable branch, and this CI test will pick up the changes).

Awesome, thanks for the work.

> By the way, I don't know if there's a better way to handle how the Prow images are updated,

Would like to have advice from @BenTheElder @Q-Lee

BenTheElder commented:
A few things.

Q-Lee has taken another job, I think we'll need to reassign review, I'm going through PRs and fixing that now...
/unassign @Q-Lee
/assign @krzyzacy 
/assign 

Catching back up on this, do we really need to run this on a VM somewhere else using docker-machine? We run nested docker on top of Prow already for various tasks, including Q-Lee's work on a docker-in-docker local cluster...

> By the way, I don't know if there's a better way to handle how the Prow images are updated, but as it is implemented now (see the Prow image config, this test will definitely fail when it's first merged until the following happens: ...

Open to suggestions on this front, but my suggestion for now is that we can probably do this PR better though by splitting PR into a) the kubetest / image changes b) the job that uses it.

We don't immediately push new images or anything quite like that because breaking EG kubetest disrupts many many jobs and merges to kubernetes. The way an update to our images looks is:
1) make the changes, have them reviewed, unit tested etc.
2) push new images (manually, by a test-infra maintainer)
3) monitor test jobs that pick up the :latest images
4) merge an update to the images used, watch for any issues

We could automate parts of 2) 3) and 4), but we really want to get 4) right to avoid breakage so having a maintainer on hand to monitor and revert is pretty necessary, and if a maintainer is on hand then we've mostly automated 2/3/4 for them (just run some script(s) and check some logs).

BenTheElder commented:
Also, thanks for all your effort getting IPv6 tested! I'll try to carve out some more time to get that moving forward. Also thanks to @MrHohn for your input.

leblancd commented:
@BenTheElder - Regarding our use of docker-machine and GCE instance vs. running on Q-Lee's docker-in-docker local cluster...  We had started development on our CI tests a while before Q-Lee's changes were merged. When we started working on this, using the Mirantis docker-in-docker scripts on a GCE instance appeared to be the best option for virtualizing the k8s network. Originally, the Mirantis DinD scripts assumed IPv4 operation, so we had to add a bunch of IPv6 support, as listed in the bullet list in this PR's description above (incl. static routes between node/pods, IPv6 CNI config, IPv6 kubeadm config, IPv6 kubelet config (for DNS), and tayga and bind9 containers for NAT64 and DNS64 functionality. I mention these IPv6 enhancements because we'd probably need to add some or all of these IPv6 enhancements to Q-Lee's docker-in-docker local cluster if/when we move to that platform. I actually considered and looked into migrating the tests to docker-in-docker local cluster when it was merged, but at that time I didn't see a good reference CI test case to start with as a baseline... that is, a CI test that was based on this DinD framework that uses Kubeadm, loads CNI reference plugins (we need bridge plugin), does E2E testing, AND includes all of the necessary log dumping from container nodes to capture for Gubernator (along with the fore-mentioned IPv6 changes to be made). It looked like about 2 months of work, and we're very anxious to get IPv6-only test coverage up, so that we can then start on dual-stack support. I think it's worth considering moving to the docker-in-docker local cluster for the future dual-stack test CI. Especially if there are some good example CIs to start with.

Thanks for the info on the image upgrade work flow! Let me know if I need to split this PR up.

BenTheElder commented:
> @BenTheElder - Regarding our use of docker-machine and GCE instance vs. running on Q-Lee's docker-in-docker local cluster... We had started development on our CI tests a while before Q-Lee's changes were merged. 

Er sorry, not that actually, though worth considering. what I mean is we support docker running inside our pods in a testing image so that tests can call `docker build` etc., so we probably don't need this to be on a GCE vm, we can probably run the mirantis dind cluster on top of our "docker daemon running inside test image" support. See EG `pull-kubernetes-cross` for a job that uses "docker in docker" not for clusters (and unrelated to Q-Lee's work) but just for building, pushing etc.

BenTheElder commented:
Oh and we do have some jobs up with Q-Lee's thing (`ci-kubernetes-dind-conformance` is one for now) and I need to discuss with folks more about mirantis dind vs the work here, but I understand the need to get ipv6 testing up, so we should move forward with that. I just want to see if we actually want/need the docker-machine on GCE bit or not. We probably don't. Our test images have docker-in-docker for docker usage while testing here:

https://github.com/kubernetes/test-infra/blob/7ed50ea69a3fa593af0d6960aad2335025718eb0/images/bootstrap/runner#L42-L50

leblancd commented:
@BenTheElder - Running the tests on GCE isn't an absolute requirement. We were thinking of moving to the local Docker-in-Docker model long term (e.g. when we test dual-stack). Up till now, I had been thinking the Q-Lee model (abandoning Mirantis DinD). I can try using the other Docker-in-Docker with Mirantis DinD, maybe at the tail end of next week (we've been trying to get a WIP doc for dual-stack out, and I'm on PTO for a few days). One concern that I have is that I had had some issues when I tried DinDinD by using Mirantis DinD with the jpetazzo/dind in IPv6 mode (bind9 and named library mismatches). There's also a small amount of rework for the log collecting Golang stuff and associated unit tests, but that's not a big deal.

dims commented:
@leblancd we are waiting for you :) https://github.com/kubernetes/cloud-provider-openstack/issues/122

fejta commented:
/lifecycle stale
What are the plans for this PR? Please rebase and/or close.

leblancd commented:
@fejta - I'd like to keep this open. I've been asked to move this from GCE-based testing to the local DinD (DinDinD), but this will take a bit of debug.

leblancd commented:
@BenTheElder - I'm just about ready to push up a new version that spins up a (multinode) cluster in a Prow container (instead of testing on GCE platform). Would you prefer me to start a new PR, or just re-use this PR? There are a lot of changes, but I wasn't sure if we would care about losing PR comment history for the GCE-based proposal.

BenTheElder commented:
I'd be fine with either another PR if we want to keep the history here, or rebasing and cleaning up. FWIW I am also following up on Q-Lee's efforts currently and I'd be interested in hearing more about ipv6 needs.

fejta commented:
/remove-lifecycle stale

BenTheElder commented on kubetest/main.go:
```diff
@@ -154,6 +156,8 @@ func defineFlags() *options {
 	flag.StringVar(&o.gcpImageProject, "image-project", "", "Project containing node image family, required when --gcp-node-image=CUSTOM")
 	flag.StringVar(&o.gcpNodes, "gcp-nodes", "", "(--provider=gce only) Number of nodes to create.")
 	flag.StringVar(&o.gcpNodeSize, "gcp-node-size", "", "(--provider=gce only) Size of nodes to create (e.g n1-standard-1).")
+	flag.Var(&o.ginkgoParallel, "ginkgo-parallel", fmt.Sprintf("Run Ginkgo tests in parallel, default %d runners. Use --ginkgo-parallel=N to specify an exact count.", defaultGinkgoParallel))
```
this change seems unrelated?

leblancd replied:
This is just cleanup, unrelated to my changes. I moved this ginkgoParallel option because it had been out-of-order alphabetically. Of course, there seem to be other options that are out of order below.

BenTheElder commented on kubetest/main.go:
```diff
@@ -168,6 +172,7 @@ func defineFlags() *options {
 	flag.StringVar(&o.nodeTestArgs, "node-test-args", "", "Test args specifically for node e2e tests.")
 	flag.BoolVar(&o.noAllowDup, "no-allow-dup", false, "if set --allow-dup will not be passed to push-build and --stage will error if the build already exists on the gcs path")
 	flag.BoolVar(&o.nodeTests, "node-tests", false, "If true, run node-e2e tests.")
+	flag.StringVar(&o.numNodes, "num-nodes", "4", "Number of nodes to create.")
```
can we make it clear what this flag is for? kubetest has too many flags :|

leblancd replied:
Maybe it would be better to eliminate this flag to keep things simple? This num-nodes option sets the number of worker nodes to spin up in the cluster. For our network connectivity testing, 2 worker nodes is sufficient. The only reason to keep it would be if other, future multi-node test cases need to have this configurable.

BenTheElder commented on prow/config.yaml:
```diff
@@ -1548,6 +1548,39 @@ presubmits:
+        - "--repo=k8s.io/$(REPO_NAME)=$(PULL_REFS)"
+        - "--repo=github.com/leblancd/kubeadm-dind-cluster=dockernet-172.20.0.0"
+        - "--upload=gs://kubernetes-jenkins/pr-logs"
+        - "--git-cache=/root/.cache/git"
```
this flag should no longer be used

BenTheElder commented on prow/config.yaml:
```diff
@@ -1548,6 +1548,39 @@ presubmits:
+      preset-service-account: "true"
+    spec:
+      containers:
+      - image: diverdane/e2e-kubeadm-dind:v20180814-829bf6a-dirty
```
can we not use a dirty image?

BenTheElder commented on prow/config.yaml:
```diff
@@ -4032,6 +4065,28 @@ periodics:
+    - image: diverdane/e2e-kubeadm-dind:v20180814-829bf6a-dirty
+      args:
+      - "--repo=k8s.io/kubernetes=master"
+      - "--repo=github.com/leblancd/kubeadm-dind-cluster=dockernet-172.20.0.0"
```
upstream does not work?

leblancd replied:
I added a note in the PR description above. We had been using the upstream stable branch for the originally proposed GCE-based CI, but there were 3 problems we encountered when running this in a DinDinD configuration (in a Prow container). We have fixes proposed to K-D-C, but we're waiting for the fixes to get merged into master, and then have the stable branch updated to inherit the changes. The fixes are:
      * [K-D-C PR # 175](https://github.com/kubernetes-sigs/kubeadm-dind-cluster/pull/175)
      * [K-D-C PR # 185](https://github.com/kubernetes-sigs/kubeadm-dind-cluster/pull/185)
      * [K-D-C PR # 199](https://github.com/kubernetes-sigs/kubeadm-dind-cluster/pull/199)

BenTheElder commented on prow/config.yaml:
```diff
@@ -4032,6 +4065,28 @@ periodics:
+      - "--repo=k8s.io/kubernetes=master"
+      - "--repo=github.com/leblancd/kubeadm-dind-cluster=dockernet-172.20.0.0"
+      - "--upload=gs://kubernetes-jenkins/logs"
+      - "--git-cache=/root/.cache/git"
```
same here, this flag should not be used

BenTheElder commented on prow/config.yaml:
```diff
@@ -4032,6 +4065,28 @@ periodics:
+      - "--git-cache=/root/.cache/git"
+      - "--timeout=75"
+      - "--clean"
+      # Privileged mode required to enable IPv6 and Docker-in-Docker mode
```
er, are we changing networking flags on the host vs just container local routing? we should avoid that. In that case we may need to do this on a VM instead :( :man_facepalming:

leblancd replied:
The networking flag net.ipv6.conf.all.disable_ipv6 will only get set to 0 in the container, not on the host.

BenTheElder replied:
:sweat_smile: thanks, SGTM

BenTheElder commented on kubetest/kubeadmdind/kubeadm_dind.go:
```diff
@@ -0,0 +1,527 @@
+	nodeLogDir = "/var/log"
+
+	// Path to Kubernetes source tree
+	kubernetesDir = "/workspace/k8s.io/kubernetes"
```
this seems very brittle

leblancd replied:
Is there a better way to point to the pulled Kubernetes repo? This is supposed to point to the code pulled by this config in the prow/config.yaml:
"--repo=k8s.io/$(REPO_NAME)=$(PULL_REFS)"

BenTheElder commented on kubetest/kubeadmdind/kubeadm_dind.go:
```diff
@@ -0,0 +1,527 @@
+	kubernetesDir = "/workspace/k8s.io/kubernetes"
+
+	// kubeadm-dind-cluster script
+	kdcScript = "/workspace/github.com/leblancd/kubeadm-dind-cluster/fixed/dind-cluster-stable.sh"
```
ditto, this seems very brittle :/

leblancd replied:
Is there a better way to point to the pulled K-D-C repo? This is supposed to point to the code pulled by this config in the prow/config.yaml:
"--repo=github.com/leblancd/kubeadm-dind-cluster=dockernet-172.20.0.0"
(Note: This URL will eventually be replaced with github.com/kubernetes-sigs/kubeadm-dind-cluster=stable")

BenTheElder replied:
kubetest is supposed to work locally as well as in CI, even ./../$REPO/$SCRIPT would probably be better.

BenTheElder commented on kubetest/kubeadmdind/kubeadm_dind.go:
```diff
@@ -0,0 +1,527 @@
+	kdcScript = "/workspace/github.com/leblancd/kubeadm-dind-cluster/fixed/dind-cluster-stable.sh"
+
+	// Maximum number of nodes to create
+	maxNumNodes = 10
```
why?

leblancd replied:
This 'maxNumNodes' can be eliminated if I remove the 'numNodes' kubetest option and use a fixed number of worker nodes for this test. (This was a sanity limit check on the numNodes config setting... it could have been less restrictive).

leblancd commented:
@BenTheElder - I modified the way that we were looking for the kubernetes tree and the K-D-C scripts per your suggestion. Should be ready for next round of review, if/when you have a chance.

BenTheElder commented on images/kubeadm-dind/kubeadm_dind_runner:
```diff
@@ -0,0 +1,27 @@
+set -o pipefail
+
+# Enable IPv6 for Kubeadm DinD Cluster based CI
+sysctl -w net.ipv6.conf.all.disable_ipv6=0
```
hmm do we need another image just to do this?

leblancd replied:
If it's okay to enable IPv6 in kubekins-e2e, then we should be able to use that image directly.

BenTheElder replied:
What if we conditionally enable it in kubekins-e2e instead?

leblancd replied:
This would be enabled based upon an environment variable that we'd set in prow/config.yaml? (That is, rather than it being based upon the --ip-mode=ipv6 config in jobs/config.json... since that flag would have to be interpreted in bootstrap.py or scenario/kubernetes_e2e.py). Something like:
      env:                                              
       - name: IPV6_ENABLE
         value: "true"

leblancd replied:
@BenTheElder - How does this look: PR #9102 ?
I figured this IPv6-enable change might make sense to do separately so that bootstrap and kubekins-e2e containers using it can be built and pushed ahead of this PR 7529.

leblancd replied:
@BenTheElder - the suggestion for PR #9102 is to move this enabling of IPv6 to kubetest (rather than offer it in the bootstrap runner, since bootstrap container is generic). Optionally, I can move this to kubernetes_e2e.py or to the kubeadm_dind deployment (difference being how generally available the IPv6 enabling is to other CI test cases). Do you have a preference where I should put this?

BenTheElder commented on prow/config.yaml:
```diff
@@ -1492,6 +1492,38 @@ presubmits:
+    rerun_command: "/test pull-kubernetes-e2e-kubeadm-dind-ipv6"
+    trigger: "(?m)^/test pull-kubernetes-e2e-kubeadm-dind-ipv6,?(\\s+|$)"
+    skip_branches:
+    - release-1.6 # doesn't support IPv6
```
we no longer support testing these branches (< 1.8) at all so we can probably drop this. we only support the officially supported versions + under development version (1.9, 1.10, 1.11; master/1.12)

leblancd replied:
Will remove the skip_branches.

leblancd commented:
@BenTheElder - I moved the (conditional) enabling of IPv6 to kubetest (per suggestion in PR #9102), removed the custom prow container (switched to using kubekins-e2e container directly), and rebased. Should be ready for the next "look-see".

BenTheElder commented:
Sorry I've gotten behind on PR reviews 😞  Taking another look today.

BenTheElder commented on jobs/config.json:
```diff
@@ -869,6 +869,22 @@
       "sig-instrumentation"
     ]
   },
+  "ci-kubernetes-e2e-kubeadm-dind-ipv6": {
```
we are inlining these into the prowjob definitions now and deprecating this file
args go after a bare `--` arg to bootstrap in the spec's args, like:
```
args:
- --repo=blah
- --
- --cluster=
- --deployment=kubeadm-dind
...
```
The scenario value goes prior to the `--` like `--scenario=kubernetes_e2e`

BenTheElder commented on jobs/config.json:
```diff
@@ -869,6 +869,22 @@
+    ],
+    "scenario": "kubernetes_e2e",
+    "sigOwners": [
+      "sig-network"
```
and ownership is managed by putting the prowjob definition in `config/jobs/kubernetes/some-sig/some-yaml.yaml`

BenTheElder commented on prow/config.yaml:
```diff
@@ -1266,6 +1266,33 @@ presubmits:
+  - name: pull-kubernetes-e2e-kubeadm-dind-ipv6
+    context: pull-kubernetes-e2e-kubeadm-dind-ipv6
+    max_concurrency: 8
+    skip_report: true
```
please also set `always_run: false`, new / unproven jobs should run by manual request

BenTheElder commented on prow/config.yaml:
```diff
@@ -2373,6 +2400,26 @@ periodics:
       image: gcr.io/k8s-testimages/kubekins-e2e:v20180730-8b7ab3104-master
 
+- interval: 6h
+  name: ci-kubernetes-e2e-kubeadm-dind-ipv6
```
let's just add the presubmit or the CI job while we work on this. with a presubmit you can request it manually on any schedule on a dummy pr like 46662, vs this one will run every 6 hours and be broken initially until we sort it out. we shouldn't add both immediately

BenTheElder commented on jobs/config.json:
```diff
@@ -869,6 +869,22 @@
+      "--extract=ci/latest-bazel",
+      "--gcp-zone=us-central1-f",
+      "--ip-mode=ipv6",
+      "--provider=local",
```
is this the right provider?

BenTheElder commented on jobs/config.json:
```diff
@@ -869,6 +869,22 @@
+      "--cluster=",
+      "--deployment=kubeadm-dind",
+      "--extract=ci/latest-bazel",
+      "--gcp-zone=us-central1-f",
```
hmm why do we still have gcp zone?

BenTheElder commented on scenarios/kubernetes_e2e.py:
```diff
@@ -572,7 +572,7 @@ def main(args):
     elif args.deployment == 'kops' and args.provider == 'gce':
         set_up_kops_gce(mode.workspace, args, mode, cluster, runner_args)
-    elif args.gce_ssh:
+    elif args.gce_ssh and args.provider != 'local':
```
this change doesn't seem right

leblancd replied:
I'm up for suggestions on how best to address this. What's happening here is that args.gce_ssh seems to be defaulted if it's not set explicitly, so the existing "elif args.gce_ssh" is always true, and an attempt is made to add GCE SSH keys. One possible way to fix this would be to change the conditional to "elif args.provider == 'gce' or args.provider == 'gcp'", but I don't know if that's going to break some kubernetes-anywhere tests. Or if there's a way to NOT default args.gce_ssh without breaking other tests? In my case (and any other CI that uses kubeadm_e2e in a local or dind mode), I want it to skip setting up the SSH keys.

krzyzacy replied:
cc @dims for this change

BenTheElder commented:
/lint

BenTheElder commented:
@krzyzacy @amwat can you help me take a look at the kubetest changes?

leblancd commented:
@krzyzacy or @amwat, any chance you can look at this? We're eager to get IPv6 CI testing in place.

krzyzacy commented on config/jobs/kubernetes/sig-network/sig-network-misc.yaml:
```diff
@@ -17,6 +17,44 @@ presubmits:
+      preset-service-account: "true"
+    spec:
+      containers:
+      - image: gcr.io/k8s-testimages/kubekins-e2e:v20180730-8b7ab3104-master
```
the image tag seems out-dated

BenTheElder commented:
Potentially [`kind`](https://sigs.k8s.io/kind) is an avenue, yes. I'd like to mature it a little bit more and I'm still not fully clear on what changes ipv6 need just yet. @pmichali spoke with me about it a bit

leblancd commented:
/area ipv6

BenTheElder commented on config/jobs/kubernetes/sig-network/sig-network-misc.yaml:
```diff
@@ -17,6 +17,47 @@ presubmits:
+      preset-service-account: "true"
+    spec:
+      containers:
+      - image: diverdane/kubekins-e2e:v20180924-703d9d3-master
```
our image bump script (experiment/bump_e2e_image.sh) is going to break this on the next image cut

leblancd replied:
@BenTheElder Is the bump script triggered automatically, or is it something that's manually triggered from time to time, or when needed? I'm just trying to minimize the turnaround time to get this running after the first merge.

BenTheElder replied:
It's manual, but right now for example it's relatively frequent as we're starting to merge changes to kubetest etc. again post code-freeze.

BenTheElder commented:
> `DIND_STORAGE_DRIVER=vfs`

Won't this have significant performance loss? Why do we need to do this ..? (if it's to avoid overlay stacking, that's possible without doing this - just use a volume)

/unassign
I'm working on stabilizing a solution built in go with our CI in mind from the get-go with a fair bit of my time these days, I think @krzyzacy and others should finish reviewing this. I don't want to be in the way but I have plenty of nits left .. I should probably leave it to someone else to approve. I've been poking others to review.

Let me know if I can answer questions though or help with our `images/bootstrap/` dind setup. I'm also interested in making ipv6 work in my tool in the nearish future :-)

leblancd commented on kubetest/kubeadmdind/kubeadm_dind.go:
```diff
@@ -0,0 +1,559 @@
+
+	// Relative path to Kubernetes source tree
+	kubeOrg  = "k8s.io"
+	kubeRepo = "kubernetes"
```
Does the go/build search only search for the kubernetes tree under the $GOPATH directory? The findPath() function below is intended to add robustness in handling different ways that kubetest might be invoked (e.g. manually with Kubernetes tree in $GOPATH, in a CI where the Kubernetes tree is in /workspace [I think that's where the Prow job repo directive places the tree], or manually where the Kubernetes tree is in the current directory, etc.). It looks for Kubernetes in /workspace, $GOPATH, and current directory. Is there a way to use go/build to search these various locations?

leblancd commented on kubetest/kubeadmdind/kubeadm_dind.go:
```diff
@@ -0,0 +1,559 @@
+		"BUILD_KUBEADM":               "y",
+		"BUILD_HYPERKUBE":             "y",
+		"IP_MODE":                     ipMode,
+		"KUBERNETES_CONFORMANCE_TEST": "yes",
```
I'll change this to match the docs.

leblancd commented:
@BenTheElder - The kubeadm-dind-script does use a volume for docker data. The following is from the host bootstrap/DinD container:
```
root@356916e09284:/workspace# docker volume ls
DRIVER              VOLUME NAME
local               kubeadm-dind-kube-master
local               kubeadm-dind-kube-node-1
local               kubeadm-dind-kube-node-2
root@356916e09284:/workspace# docker volume inspect kubeadm-dind-kube-master
[
    {
        "CreatedAt": "2018-09-27T01:55:18Z",
        "Driver": "local",
        "Labels": {
            "mirantis.kubeadm_dind_cluster_runtime": ""
        },
        "Mountpoint": "/docker-graph/volumes/kubeadm-dind-kube-master/_data",
        "Name": "kubeadm-dind-kube-master",
        "Options": {},
        "Scope": "local"
    }
]
_data356916e09284:/workspace# ls /docker-graph/volumes/kubeadm-dind-kube-master/_data/docker
containers  image  network  plugins  swarm  tmp  trust	vfs  volumes
root@356916e09284:/workspace# 
```
The setting of $DIND_STORAGE_DRIVER to vfs is done to override K-D-C's default overlay2 driver with something that isn't an overlay. I could potentially propose a change in K-D-C to allow for the storage driver to be unspecified when dockerd is started, but dockerd will default to the vfs storage driver anyway.

leblancd commented:
@krzyzacy - I've split this PR into two (as requested by @BenTheElder), one to add a K-D-C based kubetest deployer, and a followup (PR #9640) to add prow job config for the IPv6-only CI. Can you take a look when you get a chance? This change is a high priority for SIG-network, so I appreciate any attention you can give to this.

leblancd commented:
@BenTheElder - Let me know if I can help when you start looking into adding IPv6 support for KinD! The changes needed are summarized in the description above, and you can find full details here:
https://github.com/leblancd/kube-v6
The notes should be fairly self expanatory, but let me know if you have any questions.

leblancd commented:
@amwat - Do you have time to review this, or is there anyone else who has time to review this? We'd appreciate any support you could lend to get this going. This is a top priority for SIG-network, since we need to get test coverage for IPv6, and there are upcoming features that we'll want to make sure are IPv6-ready.

amwat commented on kubetest/kubeadmdind/kubeadm_dind.go:
```diff
@@ -0,0 +1,567 @@
+		return err
+	}
+
+	// Build e2e.test and ginkgo binaries
```
Probably move this to TestSetup()?

leblancd replied:
Makes sense. I'll move these builds of e2e/ginkgo to TestSetup, and test this locally.

amwat commented on kubetest/kubeadmdind/kubeadm_dind.go:
```diff
@@ -0,0 +1,567 @@
+// and searching for container names that begin with a specified name prefix.
+func (d *Deployer) detectNodeContainers(namePrefix string) ([]string, error) {
+	log.Printf("Looking for container names beginning with '%s'", namePrefix)
+	o, err := d.getOutput("kubectl get nodes --no-headers")
```
I'm curious to know where does it the the kubeconfig from at this point?

leblancd replied:
The $HOME/.kube/config file for kubectl gets created in the host DinD container via a 'kubectl config set-cluster ...' command. The K-D-C scripts do that e.g. here: https://github.com/kubernetes-sigs/kubeadm-dind-cluster/blob/master/dind-cluster.sh#L1091-L1093. (I think that's what you were asking).

amwat commented on kubetest/kubeadmdind/kubeadm_dind.go:
```diff
@@ -0,0 +1,567 @@
+
+// detectKubeContainers creates a list of containers (either running or
+// exited) on a master or worker node whose names contain any of a list of
+// Kubernetes system pod name substings.
```
nit: s/substings/substrings

leblancd replied:
Thanks, will fix.

leblancd commented:
@amwat , thanks for the review! I moved the build of test binaries to TestSetup() as suggested. I also made a small change to eliminate the use of the DIND_STORAGE_DRIVER environmental variable when running the K-D-C script, and therefore was able to get rid of the deployer.runWithEnv() method.
Tested locally: [results](https://k8s-gubernator.appspot.com/build/my-kubernetes-jenkins/pull/65951/pull-kubernetes-e2e-kubeadm-dind-ipv6/20181015-145511-3f288882682f9b13-273/).

krzyzacy commented on config/jobs/kubernetes/sig-network/sig-network-misc.yaml:
```diff
@@ -17,6 +17,47 @@ presubmits:
         - --
         - "true"
 
+  kubernetes/kubernetes:
```
^^ drop the config change in this PR?

krzyzacy commented on kubetest/kubeadmdind/BUILD.bazel:
```diff
@@ -0,0 +1,30 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
```
also add an OWNERS file for `kubeadmdind`

leblancd replied:
@krzyzacy - Whom should I list as reviewers/approvers, and do we need labels in the OWNERS file?

krzyzacy replied:
no labels are needed, and I would put yourself & people from your project as reviewers/approvers (don't need to put test-infra people here since we are already top-level approvers)

leblancd replied:
Will add, thanks.

krzyzacy commented:
I think we should unblock this? (I'm fine since the deployer code are within it's own package)... maybe also worth make an issue to make ipv6 work with kind if we have future plans? @BenTheElder ^^

leblancd commented on kubetest/main.go:
```diff
@@ -930,6 +935,13 @@ func prepare(o *options) error {
+	if o.ipMode == "ipv6" || o.ipMode == "dual-stack" {
+		log.Printf("Enabling IPv6")
+		if err := control.FinishRunning(exec.Command("sysctl", "-w", "net.ipv6.conf.all.disable_ipv6=0")); err != nil {
+			return err
```
@krzyzacy - Will we need ipMode for KinD deployment?

krzyzacy replied:
I don't know the design about that yet, I would hope each deployer can handle specific logic like this  rather than in main.go...

leblancd replied:
Okay, I'll move this flag to kubeadmdind.

leblancd commented:
@krzyzacy - I moved the IP mode flag to make it kubeadm-dind specific, added a UT test case for IP mode validation, and made a minor improvement to the existing UT (replaced if/else with case statement).
([Latest local test run](https://k8s-gubernator.appspot.com/build/my-kubernetes-jenkins/pull/65951/pull-kubernetes-e2e-kubeadm-dind-ipv6/20181017-171910-329dc332126f72f2-264/))

krzyzacy commented:
great! thanks!
/lint

krzyzacy commented:
/lgtm

krzyzacy commented:
@leblancd feel free to prep a PR for prowjob config, preferably a periodic job, I'll push a new kubetest image later today.

leblancd commented:
@krzyzacy - I have a [PR for a presubmit prowjob config](https://github.com/kubernetes/test-infra/pull/9640). BenTheElder mentioned in [this comment on Aug 27](https://github.com/kubernetes/test-infra/pull/7529#discussion_r213147538) that there was a way to manually trigger presubmit jobs with a fake PR (and job is configured with always-run=false, so it won't auto trigger). If manually triggering a presubmit job isn't practical, let me know, and I'll convert the prow config to a periodic job.
